File: problems/problems_21/solution.go
package problem21

import (
	"encoding/json"
	. "leetCode/golang/models"
	"log"
	"strings"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	for cur := dummy; list1 != nil || list2 != nil; cur = cur.Next {
		if list1 != nil && (list2 == nil || list2.Val >= list1.Val) {
			cur.Next = &ListNode{Val: list1.Val}
			list1 = list1.Next
		} else {
			cur.Next = &ListNode{Val: list2.Val}
			list2 = list2.Next
		}
	}
	return dummy.Next
}

func Solve(input string) interface{} {
	values := strings.Split(input, "\n")
	var list1 *ListNode
	var list2 *ListNode

	var list1IntArray []int
	if err := json.Unmarshal([]byte(values[0]), &list1IntArray); err != nil {
		log.Fatal(err)
	}
	list1 = IntArrayToLinkedList(list1IntArray)
	var list2IntArray []int
	if err := json.Unmarshal([]byte(values[1]), &list2IntArray); err != nil {
		log.Fatal(err)
	}
	list2 = IntArrayToLinkedList(list2IntArray)

	return LinkedListToIntArray(mergeTwoLists(list1, list2))
}

File: problems/problems_21/solution.ts
import {IntArrayToLinkedList,ListNode,LinkedListToIntArray} from "../../typescript/models/listnode";

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
	const dummy: ListNode = new ListNode();
	let node: ListNode = dummy;
	while (list1 !== null || list2 !== null) {
		if (list2 === null || (list1 !== null && list1.val < list2.val)) {
			node.next = list1;
			list1 = list1.next;
		} else {
			node.next = list2;
			list2 = list2.next;
		}
		node = node.next;
	}
	return dummy.next;
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const list1: ListNode | null = IntArrayToLinkedList(JSON.parse(inputValues[0]));
	const list2: ListNode | null = IntArrayToLinkedList(JSON.parse(inputValues[1]));
	return LinkedListToIntArray(mergeTwoLists(list1, list2));
}
